### PR TITLE
Use pattern matching in ComposableStatsCalculator

### PR DESCRIPTION
**Pull request by kokosing**
Use pattern matching in ComposableStatsCalculator

### CODE REVIEW

**Comment by kokosing** [comment]
The first commit comes from https://github.com/prestodb/presto/pull/8376, in case of comments to that commit, please comment in oryginal PR.

File: presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        this.matchablesByClass = ImmutableListMultimap.copyOf(matchablesByClass);
+    }
+
+    public Stream<T> getCandidates(Object object)
+    {
+        return Streams.stream(ancestors(object.getClass()))
+                .flatMap(clazz -> matchablesByClass.get(clazz).stream());
+    }
+
+    private static Iterator<Class> ancestors(Class clazz)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java]: I know it was here before.
But this method does not look good. It does not handle interfaces correctly.

More correct, but possibly less performant approach would be to implement `getCandidates` as iteration over keys of `matchablesByClass` 
and for each check if `key.isInstance(object)`

File: presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Streams;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class MatchablesStore<T extends Matchable>
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java]: I do not think we should use `<T extends Matchable>` here. It unnecessary adds class hierarchy dependency for users of `MatchableStore`. 

It should be 
```
public class MatchablesStore<T>
```

And in `Builder` we could have two methods
```
register(T matchable, Pattern pattern) {
   ....
} 
register(T matchable) {
   checkArgument(matchable instanceOf Matchable);
   register(matchable, ((Matchable)matchable).getPattern())
}

File: presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    public Stream<T> getCandidates(Object object)
+    {
+        return Streams.stream(ancestors(object.getClass()))
+                .flatMap(clazz -> matchablesByClass.get(clazz).stream());
+    }
+
+    private static Iterator<Class> ancestors(Class clazz)
+    {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java]: Also I do no like the name `MatchablesStore`. Storing matchables is not main responsibility. Matching is.
I would call it `MatchingEngine`

File: presto-main/src/main/java/com/facebook/presto/util/MoreMath.java [hunk 4]
```diff
@@ -102,7 +102,7 @@ else if (isNaN(right)) {
         }
         return max(left, right);
     }
-
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/MoreMath.java]: unrelated

**Review by losipiuk — CHANGES_REQUESTED** [review]

**Comment by losipiuk** [comment]
Travis failed (probably due to errors in base branch)

**Pull request merged**